include/internal/ktls.h

```diff
@@ -141,7 +162,67 @@ static ossl_inline int ktls_send_ctrl_message(int fd, unsigned char record_type,
     return sendmsg(fd, &msg, 0);
 }
 
-#   endif                       /* KERNEL_VERSION */
-#  endif                        /* OPENSSL_SYS_LINUX */
-# endif                         /* HEADER_INTERNAL_KTLS */
-#endif                          /* OPENSSL_NO_KTLS */
+#    define K_MIN1_RX  17
+#    if LINUX_VERSION_CODE < KERNEL_VERSION(K_MAJ, K_MIN1_RX, K_MIN2)
+
+#    ifndef PEDANTIC
+#     warning "KTLS requires Kernel Headers >= 4.17.0 for receiving"
+#     warning "Skipping Compilation of KTLS receive data path"
+#    endif
+
+static ossl_inline int ktls_read_record(int fd, void *data, size_t length)
+{
+    return -1;
+}
+
+#    else
+
+/*
+ * Receive a TLS record using the crypto_info provided in ktls_start.
+ * The kernel strips the TLS record header, IV and authentication tag,
+ * returning only the plaintext data or an error on failure.
+ * We add the TLS record header here to satisfy routines in rec_layer_s3.c
+ */
+static ossl_inline int ktls_read_record(int fd, void *data, size_t length)
+{
+    struct msghdr msg = {0};
+    int cmsg_len = sizeof(unsigned char);
+    struct cmsghdr *cmsg;
+    char buf[CMSG_SPACE(cmsg_len)];
+    struct iovec msg_iov;
+    int ret;
+    char *p = data;
+    int prepend_length = SSL3_RT_HEADER_LENGTH;
+
+    msg.msg_control = buf;
+    msg.msg_controllen = sizeof(buf);
+
+    msg_iov.iov_base = p + prepend_length;
+    msg_iov.iov_len = length - prepend_length - EVP_GCM_TLS_TAG_LEN;
+    msg.msg_iov = &msg_iov;
+    msg.msg_iovlen = 1;
+
+    ret = recvmsg(fd, &msg, 0);
+    if (ret < 0)
+        return ret;
+
+    if (msg.msg_controllen > 0) {
+        cmsg = CMSG_FIRSTHDR(&msg);
+        if (cmsg->cmsg_type == TLS_GET_RECORD_TYPE) {
+            *(p) = *((unsigned char *)CMSG_DATA(cmsg));
+            *(p + 1) = TLS1_2_VERSION_MAJOR;
+            *(p + 2) = TLS1_2_VERSION_MINOR;
+            *(p + 3) = ret >> 8 & 0xff;
+            *(p + 4) = ret & 0xff;
+            ret += prepend_length;
```

Comment:
WPACKET seems more appropriate here.

Reply:
WPACKET is internal to "/ssl", using it here in "/include/internal" is somewhat unusual. Additionally it adds some unnecessary allocations that could be avoided for its internal sub packets structure.

The code is not much more readable: (something like this)
            WPACKET wpkt;
            WPACKET_init_static_len(&wpkt, p, length, 0);
            WPACKET_put_bytes_u8(&wpkt, *((unsigned char *)CMSG_DATA(cmsg)));
            WPACKET_put_bytes_u8(&wpkt, TLS1_2_VERSION_MAJOR);
            WPACKET_put_bytes_u8(&wpkt, TLS1_2_VERSION_MINOR);
            WPACKET_put_bytes_u8(&wpkt, (ret >> 8) & 0xff);
            WPACKET_put_bytes_u8(&wpkt, ret & 0xff)

Reply:
OK. So I'll add a comment here. We are safe because the kernel returns a value shorter than the length provided to it, which is shorter than `(SIZE_MAX / 2)`.